docs/proposals/2841-flux-hpc/README.md

```diff
@@ -0,0 +1,275 @@
+
+### Alternative Considered
+
+#### A Hard-Coded fluxPolicy
+
+An alternative approach is to create a hard-coded `fluxPolicy` directly within the `MLPolicySource` struct. This would involve defining explicit fields for each Flux configuration parameter we want to expose.
+
+##### API Design for `fluxPolicy`
+
+This would involve adding a new `Flux *FluxMLPolicySource` field to the `MLPolicySource` struct and defining the `FluxMLPolicySource` with strongly-typed fields.
+
+```go
+type MLPolicySource struct {
+    Torch *TorchMLPolicySource `json:"torch,omitempty"`
+    MPI   *MPIMLPolicySource   `json:"mpi,omitempty"`
+    HPC   *HPCMLPolicySource   `json:"hpc,omitempty"`
+
+    // FluxMLPolicy defines policy only for Flux
+    Flux  *FluxMLPolicySource  `json:"flux,omitempty"`
+}
+
+
+// FluxMLPolicySource represents a Flux-specific runtime configuration 
+// Fields must be hard-coded fields.
+type FluxMLPolicySource struct {
+    // Tasks is the number of tasks for the Flux job.
+    // +optional
+    Tasks *int32 `json:"tasks,omitempty"`
+
+    // ViewImage specifies the container image for the Flux init container.
+    // +optional
+    ViewImage string `json:"viewImage,omitempty"`
+
+    // NetworkDevice is the network interface Flux should use for communication (e.g., "eth0").
+    // +optional
+    NetworkDevice string `json:"networkDevice,omitempty"`
+
+    // QueuePolicy sets the scheduling policy for the Fluxion scheduler (e.g., "fcfs").
+    // +kubebuilder:validation:Enum=fcfs;easy
+    // +optional
+    QueuePolicy string `json:"queuePolicy,omitempty"`
+
+    // Interactive specifies whether to run an interactive Flux shell.
+    // +optional
+    Interactive *bool `json:"interactive,omitempty"`
+}
+```
+
+There would likely be many more fields. The above is a subset as an example that we have implemented so far.
+
+##### Example `ClusterTrainingRuntime` with `fluxPolicy`
+
+```yaml
+apiVersion: trainer.kubeflow.org/v1alpha1
+kind: ClusterTrainingRuntime
+metadata:
+  name: hpc-flux-specific-runtime
+spec:
+  mlPolicy:
+    flux:
+      tasks: 4
+      viewImage: "ghcr.io/converged-computing/flux-view-ubuntu:tag-jammy"
+      networkDevice: "eth0"
+      queuePolicy: "fcfs"
+      interactive: false
+```
+
+While this approach offers the benefits of strong typing and explicit, self-documenting fields, there are several risks to be mitigated:
+
+1.  **Tight Coupling and API Brittleness:** This design tightly couples the Kubeflow Trainer API to the specific implementation details of Flux Framework. If Flux itself introduces a new, important configuration option, it would require a formal change to the Kubeflow Trainer API. E.g., if we make additional options to Flux a string field, that serves as a mitigation for API brittleness. Requiring each flag or option to be a hardened field would make future development harder and require a new release. This creates a high maintenance burden and slows down the adoption of new features.
```

Comment:
>it would require a formal change to the Kubeflow Trainer API.

I guess, this problem can be addressed if we introduce settings field to the FluxPolicy, which allows dynamically add more settings supported by Flux.

Additionally, I would like to understand how frequent settings are changed/modified in Flux framework, and what changes are required in TrainJob controller to support them ?

Reply:
> If Flux is always going to have such parameters, why we don't want to add them directly in the spec, like we did for MPI

Because those are specific to Flux, and may not be shared by every HPC workload manager that would use HPC policy. There is not a good way to standardize that beyond allow for the specific workload manager backend to look for what is required in the map and raise an error if it is missing.

Reply:
Resolving here - we are implementing Flux and adding parameters directly to the spec.